src/aiokatcp/server.py

```diff
@@ -372,7 +372,14 @@ async def _stop_impl(self, cancel: bool = True) -> None:
                 self._stop_task = asyncio.current_task()
                 if self._server is not None:
                     self._server.close()
-                    await self._server.wait_closed()
+                    # Workaround for race condition: we need any connections
+                    # that were started before the listening socket was closed
+                    # to be fully established, so that we can shut them down
+                    # properly, and that may take several event loop iterations.
+                    # See https://github.com/python/cpython/issues/109564 for
+                    # more information.
+                    for _ in range(10):
```

Comment:
Is this based on the four steps you identified in the bug report, plus / times a fudge factor?

Could you still `await self._server.wait_closed()` after the loop, or is it not necessary anymore?

Reply:
> Is this based on the four steps you identified in the bug report, plus / times a fudge factor?

It also involves the additional asynchronous call to `connection_made`. So I think in theory 2 iterations is sufficient, but shutdown isn't performance-critical, so I'd rather have a healthy safety margin.

> Could you still `await self._server.wait_closed()` after the loop, or is it not necessary anymore?

You could, but since it seems its purpose is to wait for all the connections to be closed, either it's going to be a no-op or it will hang (if there is still a race condition), so there is no good reason to call it.